### DISCUSSION TITLE
Roadmap of koordlet and slo-controller

### DISCUSSION THREAD

**Discussion by zwzhang0107** [koordinator-sh/koordinator#99] [General]
Let's define the "priority one" topics for the next few milestones of koordlet and slo-controller. @stormgbs @jasonliu747 @saintube

**Comment by jasonliu747** [comment]
In my onw opinion, documentation is the number one priority of this project.

**Reply by zwzhang0107** [reply]
How about show a list of needed docs in your mind, and we can divide the work

**Reply by jasonliu747** [reply]
SGTM

**Comment by zwzhang0107** [comment]
Currently, several topics are important in my opinion.

1. Diversify the abilities of Koordlet, including a series of strategies about QoS Management, and collaboration with scheduler such as topology-aware scheduling
2. Well-designed modules with framework, which can be easily extended by other developers. Should start with QoS-Manager and RuntimeHooks.
3. Introduction docs, including designs and guides.

**Comment by zwzhang0107** [comment]
Now we have v0.3 already defined as following.
- [ ] [cpu burst](https://github.com/koordinator-sh/koordinator/issues/52)
- [ ] [design doc of qos manager](https://github.com/koordinator-sh/koordinator/issues/58)
- [ ] [memory qos](https://github.com/koordinator-sh/koordinator/issues/55)
- [ ] [LLC and MBA isolation](https://github.com/koordinator-sh/koordinator/issues/56)

**Comment by zwzhang0107** [comment]
last discussion, we have settle down the following issues to do

1. [setting limits for extend-res](https://github.com/koordinator-sh/koordinator/issues/83) - v0.4
3. [improvement of memory evict](https://github.com/koordinator-sh/koordinator/issues/85) - v0.4
4. QoS manager 
    a. design - v0.4
    b. framework implement  - v0.5
    c. fist plugins implement - v0.6
    c. more plugins implement - v0.7
6. runtime hook
    a. framework implement - v0.4
    b. first plugin implement  - v0.5
    c. more plugins implement - v0.6
8. UT review
    a. cpu_suppress_test - v0.4
    b. memory_evict_test - v0.4
9. functional test scaffold  - v0.5
10. docs (your list is expecting @jasonliu747), let's divide and conquer!

**Reply by zwzhang0107** [reply]
update: 

1. [setting limits for extend-res](https://github.com/koordinator-sh/koordinator/issues/83) - v0.4
2. [improvement of memory evict](https://github.com/koordinator-sh/koordinator/issues/85) - v0.4
3. QoS manager 
    a. design - v0.5
    b. framework implement  - v0.5
    c. fist plugins implement - v0.6
    c. more plugins implement - v0.7
4. runtime hook
    a. framework implement - v0.4
    b. first plugin implement  - v0.4
    c. more plugins implement - v0.6
5. UT review
    a. cpu_suppress_test - v0.4
    b. memory_evict_test - v0.4
6. functional test scaffold  - v0.5
7. docs (your list is expecting @jasonliu747), let's divide and conquer!
8. other features
    a. cpu evict strategy - v0.4
    b. network qos - v0.4
    c. blkio qos - v0.5
    c. recommender - 0.6

**Comment by jasonliu747** [comment]
For a good project, there are several documents I think are essential:
1. Architecture
2. Quick Start Giude
3. Installing/Upgrading Koordinator
4. **Best Practices**
    - Memory QoS
    - Memory Evict
    - CPU Burst
    - CPU Supress
    - Deep Learning
    - Big Data
    - etc
6. FAQ
7. Troubleshooting

**Comment by zwzhang0107** [comment]
feature implement
- [x] Memory QoS
- [x] Memory Evict
- [x] CPU Burst
- [x] CPU Supress
- [x] LLC/MBA
- [x] cpu evict strategy 
- [ ] network qos 
- [ ] blkio qos

doc
- [ ] Memory QoS
- [ ] Memory Evict
- [ ] CPU Burst
- [ ] CPU Supress
- [ ] LLC/MBA
- [ ] cpu evict strategy 
- [ ] network qos 
- [ ] blkio qos

**Comment by zwzhang0107** [comment]
milestone v0.5:
  ○ support runtime-plugin bypass running mode (#120 )
  ○ qos-manager design (#58)
  ○ support batch-res limit != request (#83)
  ○ pqos，support llc/mba for amd (maybe doing some research first) (#216)
  ○ compatible internal design
